src/Text/Trifecta/Parser.hs

```diff
@@ -129,14 +130,23 @@ instance Semigroup a => Semigroup (Parser a) where
   (<>) = liftA2 (<>)
   {-# INLINE (<>) #-}
 
+{- In order to silence '-Wnoncanonical-monoid-instances' we'd need to
+   change the type-signature to emulate the not-yet existing superclass
+   relationship:
+
+instance (Semigroup a, Monoid a) => Monoid (Parser a) where
+  mappend = (<>)
+  mempty = pure mempty
+-}
 instance Monoid a => Monoid (Parser a) where
```

Comment:
@RyanGlScott as to future knowledge, that's where http://hackage.haskell.org/package/base-feature-macros can help by abstracting over it...

However, the warning is a different issue here. I'm not worried about the warning of future GHCs, but that we need a breaking change if we wanted to write the code forward-compatible, since

   instance (Semigroup a, Monoid a) => Monoid (Parser a) where

is less general than

   instance Monoid a => Monoid (Parser a) where

on older GHCs; it will become less general once PRE_SMP_GHC becomes true (at which point there will be a harmless warning; but otoh at that point the GHC version to use for CPP will become known)